draft-ietf-quic-transport.md

```diff
@@ -994,6 +994,14 @@ packets sent from only one local address.  An endpoint that migrates away from a
 local address SHOULD retire all connection IDs used on that address once it no
 longer plans to use that address.
 
+The endpoint can explicitly request its peer to retire connection IDs by sending
+a NEW_CONNECTION_ID frame that contains a Retire Prior To field.  The peer is
+required to retire these connection IDs in a timely manner.  On receipt of the
+acknowledgement for the packet that contained the NEW_CONNECTION_ID frame, the
+endpoint may start a 3 PTO timer, which on expiring may close the connection
+with a PROTOCOL_VIOLATION error if all the connection IDs were not retired as
+requested.
+
```

Comment:
I think this is too aggressive. A single delayed packet could randomly take down a connection and attackers could speculate in that.

Reply:
How could a single delayed packet take the connection down? Best case, the same packet that acknowledges the NEW_CONNECTION_ID also contains all the RETIRE_CONNECTION_ID frames. Worst case, maybe you send them an RTT later. The timer here is 3 PTOs, not 3 RTT. If you haven't retired in that amount of time, how long of a time window should we allow?

Reply:
The other option here is not to kill the connection, but the sender just removes those CIDs from its lookup/routing tables. This means, if the peer continues to get them, they will get a stateless reset. I prefer the explicit behavior, because I feel like it is easier to diagnose when there is a problem; but I am willing to accept either solution.

Reply:
Yes, I was thinking 3 RTT. 3 PTO may be enough, but an attacker could still delay so I prefer to drop such packets.

If you remove them from the routing table that is a deployment strategy - but most would likely keep a grace period before resetting or dropping. I'd imagine the middlebox load balancer is slow to update and only drops while backends are more aggressive because they are directly targeted. This is actually also true without explicit retirement - do we have any text on how to deal with that?

Reply:
I would also prefer a requirement that the peer MUST change CIDs immediately. I heard push back on that design. I would like to hear reasons why that wouldn't be a good idea, if anyone has some.

Reply:
Mostly because "change immediately" is a difficult requirement to police.  Couching it in more concrete terms -- error on use of this CID in a packet with a higher packet number than the one which carried the ACK -- makes it easier to police in theory, but highlights what different parts of the system have to be coordinated to do it in practice.

Frankly, I'd even be okay with no longer recognizing those CIDs 3xPTO after sending the frame.  Fewer moving parts, easy to understand, but a small risk of stateless reset if the packet is severely delayed.